core/src/main/java/com/scalar/db/transaction/consensuscommit/RollbackMutationComposer.java

```diff
@@ -84,28 +85,24 @@ public void add(Operation base, TransactionResult result) {
     }
   }
 
-  private Put composePut(Operation base, TransactionResult result) {
+  private Put composePut(Operation base, TransactionResult result) throws ExecutionException {
     assert result.getState().equals(TransactionState.PREPARED)
         || result.getState().equals(TransactionState.DELETED);
 
+    TransactionalTableMetadata metadata = tableMetadataManager.getTransactionalTableMetadata(base);
+    LinkedHashSet<String> beforeImageColumnNames = metadata.getBeforeImageColumnNames();
+
     Map<String, Value<?>> map = new HashMap<>();
     result
         .getValues()
         .forEach(
             (k, v) -> {
-              if (k.startsWith(Attribute.BEFORE_PREFIX)) {
+              if (beforeImageColumnNames.contains(k)) {
```

Comment:
This is the main change. Here, to identify the column is a before image column, we use `beforeImageColumnNames` returned by `TransactionalTableMetadata` instead of checking if the prefix is `Attribute.BEFORE_PREFIX`.